designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala

```diff
@@ -0,0 +1,534 @@
+    @derive(schema, decoder)
+    final case class UIValueParameterDto(
+        name: String,
+        typ: TypingResultInJson,
+        expression: Expression
+    )
+
+    implicit lazy val expressionSchema: Schema[Expression]           = Schema.derived
+    implicit lazy val caretPosition2dSchema: Schema[CaretPosition2d] = Schema.derived
+
+    // Request doesn't need valid encoder
+    @derive(schema, decoder)
+    final case class ExpressionSuggestionRequestDto(
+        expression: Expression,
+        caretPosition2d: CaretPosition2d,
+        variableTypes: Map[String, TypingResultInJson]
+    )
+
+    implicit val expressionSuggestionRequestDtoEncoder: Encoder[ExpressionSuggestionRequestDto] =
+      Encoder.encodeJson.contramap[ExpressionSuggestionRequestDto](_ => Json.Null)
+
+    // Response doesn't need valid decoder
+    @derive(schema, encoder)
+    final case class ExpressionSuggestionDto(
+        methodName: String,
+        refClazz: TypingResult,
+        fromClass: Boolean,
+        description: Option[String],
+        parameters: List[ParameterDto]
+    )
+
+    object ExpressionSuggestionDto {
+
+      def apply(expr: ExpressionSuggestion): ExpressionSuggestionDto = {
+        new ExpressionSuggestionDto(
+          expr.methodName,
+          expr.refClazz,
+          expr.fromClass,
+          expr.description,
+          expr.parameters.map(param => ParameterDto(param.name, param.refClazz))
+        )
+      }
+
+    }
+
+    implicit val expressionSuggestionDtoDecoder: Decoder[ExpressionSuggestionDto] =
+      Decoder.instance[ExpressionSuggestionDto](_ => ???)
+
+    // Response doesn't need valid decoder
+    @derive(schema, encoder)
+    final case class ParameterDto(
+        name: String,
+        refClazz: TypingResult
+    )
+
+    def prepareTypingResultDecoder(modelData: ModelData): Decoder[TypingResult] = {
+      new TypingResultDecoder(name =>
+        ClassUtils.forName(name, modelData.modelClassLoader.classLoader)
+      ).decodeTypingResults
+    }
+
+    def prepareTestFromParametersDecoder(modelData: ModelData): Decoder[TestFromParametersRequest] = {
+      implicit val typeDecoder: Decoder[TypingResult] = prepareTypingResultDecoder(modelData)
+      implicit val testSourceParametersDecoder: Decoder[TestSourceParameters] =
+        deriveConfiguredDecoder[TestSourceParameters]
+      deriveConfiguredDecoder[TestFromParametersRequest]
+    }
+
+    @JsonCodec(encodeOnly = true) final case class TestSourceParameters(
+        sourceId: String,
+        parameterExpressions: Map[String, Expression]
+    )
+
+    @JsonCodec(encodeOnly = true) final case class TestFromParametersRequest(
+        sourceParameters: TestSourceParameters,
+        scenarioGraph: ScenarioGraph
+    )
+
+    @JsonCodec(encodeOnly = true) final case class ParametersValidationRequest(
+        parameters: List[UIValueParameter],
+        variableTypes: Map[String, TypingResult]
+    )
+
+    object ParametersValidationRequest {
```

Comment:
I would move this logic to the HttpService because it has more functions than a simple dto.